docs/charmsindetail.md

```diff
@@ -0,0 +1,304 @@
+
+![New charm interfaces](./diagrams/charm_interfaces_drawing_new.jpg)
+
+That's enough background for now, you can move onto Charm Writing in detail.
+
+## Charm writing
+
+### Add the operator framework
+
+To begin a new charm, you will need to import the operator framework. The charm operator framework is imported by adding the framework in the 'lib' diretory and as a submodule:
+
+```
+git submodule add https://github.com/canonical/operator mod/operator
+ln -s ../mod/operator/ops lib/ops
+```
+
+### Add interface dependences
+
+To begin a new charm we should take note of an Interfaces that might be needed, the operator framework uses Interfaces as dependencies.
+
+These dependencies are pulled in as git submodules, and should be added to the `.gitmodules` file, an example file looks like this:
+
+```
+
+[submodule "mod/operator"]
+	path = mod/operator
+	url = https://github.com/canonical/operator
+[submodule "mod/interface-mysql"]
+	path = mod/interface-mysql
+	url = git@github.com:johnsca/interface-mysql.git
+[submodule "mod/interface-http"]
+	path = mod/interface-http
+	url = git@github.com:johnsca/interface-http.git
+[submodule "mod/resource-oci-image"]
+	path = mod/resource-oci-image
+	url = git@github.com:johnsca/resource-oci-image.git
+```
+
+You can then pull in those dependencies with the following commands:
+
+```
+git submodule init
+git submodule update
+```
+
+These commands will pull in the dependencies to mod, although we will be referencing from the `lib` directory. To fix this you will need to create symbolic references.
+
+```
+ln -s ./mod/interface-mysql/interface_mysql.py ./lib/interface_mysql.py
+```
+
+This is requied for all of the required submodules.
+
+> You can learn more about writing Interfaces [here](./CharmInterfaces.md)
+
+## Imports
+
+The operator charms are invoked as scripts on the running machine. As such you will be required to add a shbang at the top of the python file:
+
+```
+#!/usr/bin/env python3
+```
+
+After this, you will then need to adapt the system path to import the framework:
+
+```
+import sys
+sys.path.append('lib')
+```
+
+Once this is done, you can then move on to import the Framework classes
+
+```
+from ops.charm import CharmBase
+from ops.main import main
+from ops.framework import StoredState
+from oci_image import OCIImageResource
+```
+
+Followed by any other imports, for example, interfaces:
+
+```
+from interface_mysql import MySQLClient
+```
+
+## Creating the charm class
+
+All Operator Framework charms should inherit from the `CharmBase` class. This gives them access to the internals
+
+For example:
+
+```
+class MyCharm(CharmBase):
+```
+
+### The charm `__init__` method
+
+The charm __init__ method has the following signature:
+
+```
+def __init__(self, framework, key)
+             ^^^^  ^^^^^^^^^  ^^^
+             ||||  |||||||||  ||||
+
+ 1. Obvious!       |||||||||  ||||
```

Comment:
Would avoid the word "obvious" anywhere. It's really offputting to anyone who doesn't find things obvious.